File: 0002-add-two-numbers/0002-add-two-numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) 
{
    
    
    let preh = new ListNode(0);
    let p3 = preh,c=0;
    
    while(l1 || l2)
    {
        let a = l1 && l1.val;
        let b = l2 && l2.val;
        let sum = a+b+c;
        
        c = Math.floor(sum/10) 
        p3.next = new ListNode(sum%10);
        
        if(l1) l1=l1.next;
        if(l2) l2=l2.next;
        p3 = p3.next;
       
        
    }
    if(c) p3.next = new ListNode(c);
    
    return preh.next;
    
};

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode *reverse(ListNode *head)
    {
       if(!head || !head->next) return head;
        
        ListNode *node = reverse(head->next);
        head->next->next = head;
        head->next = NULL;
        
        return node;
    }
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) 
    {
        // l1  = reverse(l1);
        // l2  = reverse(l2);
        
        ListNode *p1 = l1, *p2 = l2;
        
        int c = 0;
        
        ListNode *preh = new ListNode(0);
        ListNode *p3 = preh;
        while(p1 && p2)
        {
            p3->next = new ListNode((p1->val+p2->val+c)%10);
            c = (p1->val+p2->val+c)/10;
            
            p1 = p1->next;
            p2 = p2->next;
            p3 = p3->next;
        }
        
        while(p1)
        {
            p3->next = new ListNode((p1->val+c)%10);
            c = (p1->val+c)/10;  
            p1 = p1->next;
            p3 = p3->next;
        }
        
        while(p2)
        {
            p3->next = new ListNode((p2->val+c)%10);
            c = (p2->val+c)/10;  
            p2 = p2->next;
            p3 = p3->next;
        }
        if(c) p3->next = new ListNode(c);
        
        return  preh->next;
    }
};